README.md
# factorio-coin-generation

A mod to allow multiple ways to generate coins in game; doing private science research.

![Private research institute](https://thumbs.gfycat.com/BaggyEarlyArabianoryx-size_restricted.gif)

Private research institute
------------

- A facility that takes in science packs and does private research (process) to produce coins. This facility is approximately equal to a science lab in terms of per science pack processing times, module options and cost.
- 2 processing types for private research of science packs:
    - Per science pack: a separate recipe for each science pack giving the packs value in coins. Recipe ingredient quantity and processing times are generated to ensure a fair coin value, rather than allowing heavy skewing due to rounding.
    - Grouped science pack (non space): a recipe requiring 1 of each science pack below space science.
    - Grouped science pack (with space): a recipe requiring 1 of each science pack including space science.
- An option to set the value decrease percentage you get from doing private research. This is to add a premium to coin generation to enable balancing with other coin generation methods and spending mods.
- Science coin value is based on the same system as my other coin related mods:
    - Automation science pack = 1.1
    - Logistic science pack = 2.7
    - Military science pack = 6
    - Chemical science pack = 10
    - Production science pack = 33.8
    - Utility science pack = 37.9
    - Space science pack = 80.4

Biter killing bounty
---------------

- A bounty is paid for the killing of the biters and their bases, with different values for players and turrets.
- The biter & worm coin value is based on enemy health and scales logarithmically to slower increase with further health gain. Small biter is ~1 coin, large is ~4 and a behemoth is ~7 coins. Partial coin values are handled by chance. You can control a multiplier on this.
- Biter nest is a static coin value.
- Option to set turret coin value as a multiplier of player in fractions, i.e. 0.1
- Options to set the coin delivery location for players and turrets separately. Including in the player inventory, on the ground by the killer, by the corpse or in other modded collection points (Prime Intergalactic Delivery payment chest).


Planned Features
------------

- Scheduling regular coin deliveries and optionally use the item delivery pod mod. Started but doesn't feel right options wise so put on hold and hidden in mod.


Muppet Coin Based Mod Collection
------------------

This mod is part of my collection of mods that use the vanilla Factorio coins. They are designed to work together or separately as required. You can also mix with other peoples mods that use vanilla Factorio coins.

- Prime Intergalactic Delivery: a market to buy items for coins.
- Item Delivery Pod: a crashing spaceship that can bring items to the map with an explosive delivery.
- Coin Generation: a mod with a variety of ways for players and streamers to generate/obtain coins.
- Streamlabs RCON Integration: an external tool that lets streamers trigger in game actions from Streamlabs: https://github.com/muppet9010/Streamlabs-Rcon-Integration

A number of my other mods are designed for streamers or have features to make them streamer friendly.


Notes
---------

- Doesn't account for expensive mode.
- Doesn't support modded science packs.

changelog.txt
---------------------------------------------------------------------------------------------------
Version: 20.0.5
Date: 2022-11-30
  Features:
    - Updated to Factorio 1.1 as-is.
  Changes:
  Bugfixes:
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.0.4
Date: 2020-03-26
  Features:
  Changes:
  Bugfixes:
    - Really fix the event registration issue this time from 18.0.3.
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.0.3
Date: 2020-03-22
  Features:
  Changes:
  Bugfixes:
    - Fix event registration issue that causes mod mismatch/unsafe save/load warnings on cross mod versions between save and playing.
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.0.2
Date: 2020-03-19
  Features:
  Changes:
  Bugfixes:
    - handle the Mining Drones mod and its odd use of "killing" enemies.
  Info:
---------------------------------------------------------------------------------------------------
Version: 18.0.1
Date: 2020-03-12
  Features:
  Changes:
  Bugfixes:
    - fixed reported crash if biters killed with no known "killer". In this case coins are dropped on floor.
    - fixed possible crash if Prime Intergalactic delivery mod failed to return a payment chest for some reason.
  Info:
    - clarified readme and mod setting descriptions around disabling features and default/recommended values.
---------------------------------------------------------------------------------------------------
Version: 18.0.0
Date: 2020-02-26
  Features:
    - Initial Release
  Changes:
  Bugfixes:
  Info:

prototypes/private-research-institute-furnace-recipes.lua
local Utils = require("utility/utils")

return function()
    local valueDecreaseSettingRaw = settings.startup["coin_generation-private_research_institute_value_decrease_percent"].value
    local valueDecreaseSettingMultiplier = (100 - valueDecreaseSettingRaw) / 100

    data:extend(
        {
            {
                type = "recipe-category",
                name = "coin_generation-private_research_institute_furnace_science"
            }
        }
    )

    local coinItemPrototype = Utils.DeepCopy(data.raw["item"]["coin"])
    local function CreateScienceRecipe(item, coinQuantity, scienceQuantity)
        local recipeName = "coin_generation-private_research_institute_" .. item.name
        data:extend(
            {
                {
                    type = "recipe",
                    name = recipeName,
                    category = "coin_generation-private_research_institute_furnace_science",
                    localised_name = {"recipe-name.coin_generation-private_research_institute_science"},
                    localised_description = {"recipe-description.coin_generation-private_research_institute_science"},
                    energy_required = 10 * scienceQuantity,
                    ingredients = {{item.name, scienceQuantity}},
                    result = "coin",
                    result_count = coinQuantity,
                    hidden = true,
                    main_product = "",
                    icons = {
                        {
                            icon = coinItemPrototype.icon,
                            icon_size = coinItemPrototype.icon_size,
                            icon_mipmaps = coinItemPrototype.icon_mipmaps
                        },
                        {
                            icon = item.icon,
                            icon_size = item.icon_size,
                            icon_mipmaps = item.icon_mipmaps,
                            scale = 0.5
                        }
                    },
                    subgroup = "science-pack"
                }
            }
        )

        table.insert(data.raw["module"]["productivity-module"].limitation, recipeName)
        table.insert(data.raw["module"]["productivity-module-2"].limitation, recipeName)
        table.insert(data.raw["module"]["productivity-module-3"].limitation, recipeName)
    end

    --table of values for the time being - modify to be dynamic when i do the same in Prime Intergalactic Delivery mod. values taken from that mod.
    local scienceValues = {
        ["automation-science-pack"] = 1.1,
        ["logistic-science-pack"] = 2.7,
        ["military-science-pack"] = 6,
        ["chemical-science-pack"] = 10,
        ["production-science-pack"] = 33.8,
        ["utility-science-pack"] = 37.9,
        ["space-science-pack"] = 80.4
    }

    --can use data.raw["lab"]["lab"].inputs when programatically calculated
    for itemName, scienceValue in pairs(scienceValues) do
        local item = Utils.DeepCopy(data.raw["tool"][itemName])
        local scienceQuantity
        scienceValue = scienceValue * valueDecreaseSettingMultiplier
        if scienceValue >= 10 then
            scienceQuantity = 1
        elseif scienceValue >= 1 then
            scienceQuantity = Utils.RoundNumberToDecimalPlaces(5 / scienceValue, 0)
        elseif scienceValue < 1 then
            scienceQuantity = Utils.RoundNumberToDecimalPlaces(1 / scienceValue, 0)
        end
        local coinQuantity = Utils.RoundNumberToDecimalPlaces(scienceQuantity * scienceValue, 0)
        CreateScienceRecipe(item, coinQuantity, scienceQuantity)
    end
end
